core/src/hir/lowering.rs

```diff
@@ -778,32 +808,374 @@ fn lower_return_type(
+                    ReturnFallability::Fallible(ok_ty, err_ty),
+                    return_ltl.finish(),
+                )),
                 _ => Err(()),
             }
-        }
-        ast::TypeName::Unit => {
+        }),
+        ast::TypeName::Unit => return_ltl.map(|return_ltl| {
             if takes_writeable {
-                Ok(ReturnFallability::Infallible(Some(ReturnType::Writeable)))
+                (
+                    ReturnFallability::Infallible(Some(ReturnType::Writeable)),
+                    return_ltl.finish(),
+                )
             } else {
-                Ok(ReturnFallability::Infallible(None))
+                (ReturnFallability::Infallible(None), return_ltl.finish())
+            }
+        }),
+        _ => return_ltl.and_then(|mut return_ltl| {
+            lower_out_type(
+                return_type,
+                Ok(&mut return_ltl),
+                lookup_id,
+                in_path,
+                env,
+                errors,
+            )
+            .map(|ty| {
+                (
+                    ReturnFallability::Infallible(Some(ReturnType::OutType(ty))),
+                    return_ltl.finish(),
+                )
+            })
+        }),
+    }
+}
+
+pub mod elision {
+    use super::*;
+    use crate::hir::{ExplicitLifetime, ImplicitLifetime, LifetimeNode, TypeLifetime};
+
+    /// Generator for unique [`ImplicitLifetime`]s.
+    pub(in crate::hir) struct ImplicitLifetimeGenerator {
+        next: u32,
+    }
+
+    impl ImplicitLifetimeGenerator {
+        /// Returns a new [`ImplicitLifetimeGenerator`].
+        pub fn new() -> Self {
+            Self { next: 1 }
+        }
+
+        /// Returns the next [`ImplicitLifetime`].
+        pub fn gen(&mut self) -> ImplicitLifetime {
+            let label = self.next;
+            self.next += 1;
+            ImplicitLifetime::new(label)
+        }
+    }
+
+    pub(super) struct BaseLifetimeLowerer<'ast> {
+        elided_node_gen: ImplicitLifetimeGenerator,
+        method_lifetime_env: &'ast ast::LifetimeEnv,
+        nodes: SmallVec<[LifetimeNode; 2]>,
+    }
+
+    #[derive(Copy, Clone)]
+    enum ElisionSource {
+        /// No borrows in the input, no elision.
+        NoBorrows,
+        /// `&self` or `&mut self`, elision allowed.
+        SelfParam(TypeLifetime),
+        /// One param contains a borrow, elision allowed.
+        OneParam(TypeLifetime),
+        /// Multiple borrows and no self borrow, no elision.
+        MultipleBorrows,
+    }
+
+    /// The first phase of output elision inference.
+    ///
+    /// In the first phase, the type signature of the `&self` or `&mut self` type
+    /// is lowered into its HIR representation, if present. According to elision
+    /// rules, this reference has the highest precedence as the lifetime that
+    /// goes into elision in the output, and so it's checked first.
+    pub(super) struct SelfParamLifetimeLowerer<'ast> {
```

Comment:
Different states have different meanings. For example, the `SelfParamLifetimeLowerer` type is only used on the self type, since the self type is special in elision inference. There's only one self type (or none), so it's nice to have functions the consume ownership of the type, proving at the type level it's only called once. Same with the second state, which has the potential to set the lifetime that elision in the output refers to, whereas the last state can only read that state. They all lower lifetimes, but they represent different stages in elision inference, so it's nice to have a state machine.